src/server/game/Server/WorldSession.cpp

```diff
@@ -847,8 +847,20 @@ void WorldSession::ReadMovementInfo(WorldPacket &data, MovementInfo* mi)
         in conjunction with any of the moving movement flags such as MOVEMENTFLAG_FORWARD.
         It will freeze clients that receive this player's movement info.
     */
-    REMOVE_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_ROOT),
-        MOVEMENTFLAG_ROOT);
+    if (!GetPlayer()->GetVehicleBase())
+    {
+        // Only remove here for non vehicles
+        REMOVE_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_ROOT),
+            MOVEMENTFLAG_ROOT);
+    }
+    else
+    {
+        // Actually players in rooted vehicles still send commands, don't clear root for these!
+        // Check for non movement flag. Otherwise, also clear for vehicles
+        if (!GetPlayer()->GetVehicleBase()->HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_DISABLE_MOVE))
+            REMOVE_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_ROOT),
+                MOVEMENTFLAG_ROOT);
```

Comment:
This is the point though. If you put debug logs here, you'll see it fires over and over when the user is aiming the turret in a cannon for example. This point I've raised here already and the comment makes clear the intention.

So, perhaps in most cases the client wouldn't send commands when a player is rooted. But for rooted vehicles it does, and this falls foul of this code, which releases the root.

Hence, it specifically makes an exception if on a vehicle. It's the crux of this PR and I cannot see any other way of solving the issue on a core level (without making various hacky tricks elsewhere).